src/reader/mod.rs

```diff
@@ -1930,3 +1933,68 @@ mod test {
     pub(super) use check;
     pub(super) use small_buffers;
 }
+
+fn read_to_string_impl<R: BufRead>(starting_buffer_position: usize, reader: &mut R, parser: Parser, end: QName, buf: &mut Vec<u8>) -> Result<Vec<u8>> {
+    // Initialize a new XMLParser
+    let reader = StoredReader::new(
+        reader, starting_buffer_position
+    );
+    let mut temp_reader: Reader<StoredReader<&mut R>> = Reader {
+        reader,
+        parser
+    };
+
+    let range = temp_reader.read_to_end_into(end, buf)?;
+    Ok(temp_reader.into_inner().get_stored_value(range.start, range.end))
+}
+
+/// Wraps an existing Reader and stores all the bytes read out of the `inner`.
+struct StoredReader<R> {
+    inner: R,
+    initial_offset: usize,
+    stored: Vec<u8>
+}
+
+impl<R: Read> Read for StoredReader<R> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let result = self.inner.read(buf);
+        if let Ok(size) = &result {
+            self.stored.extend_from_slice(&buf[0..*size]);
+        }
+        result
+    }
+}
+
+impl<R> StoredReader<R> {
+    fn new(inner: R, initial_offset: usize) -> Self {
+        Self {
+            inner, initial_offset, stored: Vec::new()
+        }
+    }
+    fn get_stored_value(mut self, start: usize, end: usize) -> Vec<u8> {
+        // Pray that the start is 0 so we don't have to allocate
+        let start = start - self.initial_offset;
+        let end = end - self.initial_offset;
+
+        if start == 0 {
+            self.stored.truncate(end);
+            self.stored
+        } else {
+            unimplemented!()
```

Comment:
That macros should not be used in the final code. If you have unreachable branches, you should use `unreachable!` macro and is OBLIGATORY mark it with the `// SAFETY` comment describing, why it is safe to call it